src/lsm/manifest.zig

```diff
@@ -204,22 +219,29 @@ pub fn ManifestType(comptime Table: type, comptime Storage: type) type {
             manifest.manifest_log.insert(log_level, table);
 
             if (constants.verify) {
-                assert(manifest_level.contains(table));
+                assert(manifest_level.contains(.{ .copy = table }));
             }
         }
 
-        /// Updates the snapshot_max on the provide table for the given level.
-        /// The table provided is mutable to allow its snapshot_max to be updated.
-        pub fn update_table(
-            manifest: *Manifest,
-            level: u8,
-            snapshot: u64,
-            table: *TableInfo,
-        ) void {
+        /// Updates the snapshot_max on the provided table for the given level.
+        /// The table provided could either be a pointer to the table (from_level) in the
+        /// ManifestLevel or a pointer to a copy of that table (copy).
+        pub fn update_table(manifest: *Manifest, level: u8, snapshot: u64, table_union: Level.TableInfoPtr) void {
```

Comment:
Afaict we only use the `copy` branch from a test. Can we drop it and just always pass in an internal reference? `struct { *TableInfo, generation }`.

Reply:
Yes, [ManifestLevel.set_snapshot_max](https://github.com/tigerbeetledb/tigerbeetle/blob/ef8b53e44d3b93176b2d7bf257bb2f8637ed06e8/src/lsm/manifest_level.zig#L162) is only invoked with an `external TableInfoReference` via [this](https://github.com/tigerbeetledb/tigerbeetle/blob/ef8b53e44d3b93176b2d7bf257bb2f8637ed06e8/src/lsm/manifest_level.zig#L973) ManifestLevel test! 

Are you recommending that we allow updating a table only using an `internal TableInfoReference`? Specifically, are you suggesting we `assert(table_ref == .internal)` in ManifestLevel.set_snapshot_max & get rid of [this](https://github.com/tigerbeetledb/tigerbeetle/blob/ef8b53e44d3b93176b2d7bf257bb2f8637ed06e8/src/lsm/manifest_level.zig#L175-L194) piece of code for updating via `external TableInfoReference`?

Reply:
A rule of thumb I use is to narrow the types as much as possible.

So, for example
- Right now `insert_table` takes a `TableInfoReference` which is always external. So instead it should just take a `*TableInfo`.
- `LeastOverlapTable` and `OverlapRange` contain `TableInfoReference`s. But those references are always internal. So it should just contain `struct { table_info: *const TableInfo, generation: u64 }` instead.
- `set_snapshot_max` is always called internal (ignoring the test, which can be modified). So it should just receive a `struct { table_info: *const TableInfo, generation: u64 }` instead. 
- And so forth. I did a quick look through and it seemed like no operations actually need to handle both internal and external, always one or the other.

If no operations actually need to handle both internal _and_ external pointers, then we don't need to `union(enum)`s at all. I would create a `struct { table_info: *const TableInfo, generation: u64 }` (e.g. named `TableInfoReference` since that is now free) to use for internal tables, and just use `*const TableInfo` for external.

Reply:
Thank you for the advice, @sentientwaffle! 

So as of now, these functions need to handle both internal & external references:

* `remove_table` - [Manifest.remove_invisible_tables](https://github.com/tigerbeetledb/tigerbeetle/blob/ef8b53e44d3b93176b2d7bf257bb2f8637ed06e8/src/lsm/manifest.zig#L324) uses internal references & [Manifest.move_table](https://github.com/tigerbeetledb/tigerbeetle/blob/ef8b53e44d3b93176b2d7bf257bb2f8637ed06e8/src/lsm/manifest.zig#L285) uses external references. But there is no difference in the behaviour of `remove_table` based on this, except we can only  `assert(!level.contains(table))` in `remove_table` if the removal is done via an external reference (since an internal reference is invalidated). 

*  `set_snapshot_max` 

`insert_table` only needs to handle external references.

If we're change [this](https://github.com/tigerbeetledb/tigerbeetle/blob/ef8b53e44d3b93176b2d7bf257bb2f8637ed06e8/src/lsm/manifest_level.zig#L973) ManifestLevel test to use internal references, then I do agree that exposing `TableInfoReference` & `TableInfoReferenceConst` would be overkill just for `remove_table`! 

I can revert to a simpler version where:

- `TableInfoReference` is defined as `struct { table_info: *TableInfo, generation: u64 }` instead of an `enum`
-  `Manifest.update_table` & `ManifestLevel.set_snapshot_max` take `TableInfoReference` as argument since they only work with internal references
- `LeastOverlapTable` & `TableOverlapRange` uses `TableInfoReference`
- `ManifestLevel.insert_table` & `ManifestLevel.remove_table` take `const* TableInfo` as input, like they did before
- `ManifestLevel.contains` currently contains our logic for ensuring that internal pointers are valid, I can keep that logic within `Manifest.update_table` & `ManifestLevel.set_snapshot_max` since they're the only functions that work with internal pointers and need that validation: https://github.com/tigerbeetledb/tigerbeetle/blob/ef8b53e44d3b93176b2d7bf257bb2f8637ed06e8/src/lsm/manifest_level.zig#L484-L492

Does the aforementioned design look okay?

Reply:
> Does the aforementioned design look okay?

Yeah, looks good.

> So as of now, these functions need to handle both internal & external references:
> - `set_snapshot_max`

Where does the external `set_snapshot_max` call originate from? I missed it earlier.

Reply:
Ah right, I discounted that since I figured it was worth just changing the test if necessary. Thanks!

Reply:
@sentientwaffle: FYI, I've implemented the changes that we discussed in this thread as part of https://github.com/tigerbeetledb/tigerbeetle/pull/875/commits/2c745c71bf24b14ee4120dd0dc9e7df690f74210!